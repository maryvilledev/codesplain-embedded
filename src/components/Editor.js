import React, { Component } from 'react';
import CodeMirror from 'react-codemirror';

import 'codemirror/lib/codemirror.css';

const codeMirrorOptions = {
  lineNumbers: true,
  lineWrapping: true,
};

class Editor extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <CodeMirror
        options={codeMirrorOptions}
        value={this.props.snippet}
      />
    );
  }
}

export default Editor;
