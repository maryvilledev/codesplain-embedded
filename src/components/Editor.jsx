import React, { Component } from 'react';
import CodeMirror from 'react-codemirror';

import 'codemirror/lib/codemirror.css';

import highlight, {styleAll, styleLine} from '../util/highlight';

const codeMirrorOptions = {
  lineNumbers: true,
  lineWrapping: true,
  gutters: ['annotations', 'CodeMirror-linenumbers'],
  readOnly: true,
};

const makeMarker = () => {
  const marker = document.createElement('div');
  marker.style.color = '#822';
  marker.innerHTML = '●';
  return marker;
};

const pushValueToCodeMirror = (value, codeMirror) => {
  if (codeMirror.getValue() === '' && value !== '') {
    codeMirror.setValue(value);
  }
};

class Editor extends Component {
  constructor(props) {
    super(props);

    this.handleGutterClick = this.handleGutterClick.bind(this);
    this.emphasizeLine = this.emphasizeLine.bind(this);
    this.deEmphasize = this.deEmphasize.bind(this);
  }

  componentDidMount() {
    const codeMirror = this.codeMirror.getCodeMirror();
    codeMirror.on('gutterClick', this.handleGutterClick);
  }

  componentDidUpdate() {
    const {
      markedLines,
      AST,
      filters,
      snippet,
      language,
      openLine
    } = this.props;

    const codeMirror = this.codeMirror.getCodeMirror();

    pushValueToCodeMirror(snippet, codeMirror);
    this.deEmphasize(codeMirror);
    if (openLine !== undefined) { //Can't just say openLine because javascript
                                  //thinks 0 is falsey
      this.emphasizeLine(codeMirror, openLine)
    }
    markedLines.forEach((lineNumber) => {
      codeMirror.setGutterMarker(Number(lineNumber), 'annotations', makeMarker());
    });
    highlight(codeMirror, AST, filters, language);
  }

  handleGutterClick(instance, lineNumber) {
    const { onGutterClick } = this.props;
    const lineText = instance.getLine(lineNumber);
    onGutterClick(lineNumber, lineText);
  }

  emphasizeLine(codeMirror, line) {
    // Fade out the background
    const backgroundCSS = 'opacity: 0.5; font-weight: normal;';
    styleAll(codeMirror, backgroundCSS);

    // Bold the passed-in line
    const foregroundCSS = 'font-weight: bold; opacity: 1.0;';
    styleLine(codeMirror, line, foregroundCSS);
  }
  deEmphasize(codeMirror) {
    const css = 'font-weight: normal; opacity: 1.0;';
    styleAll(codeMirror, css);
  }

  render() {
    return (
      <CodeMirror
        options={codeMirrorOptions}
        value={this.props.snippet}
        ref={cm => this.codeMirror = cm}
      />
    );
  }
}

export default Editor;
