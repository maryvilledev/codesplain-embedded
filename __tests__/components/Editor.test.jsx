import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json'

import Editor from '../../src/components/Editor';

describe('<Editor />', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(
      <Editor
        markedLines={[0]}
        AST={{}}
        filters={{}}
        snippet="Wubba lubba Dub Dub"
        language="bad"
        openLine={0}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
