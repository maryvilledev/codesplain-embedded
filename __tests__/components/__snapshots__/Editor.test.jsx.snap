// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Editor /> matches snapshot 1`] = `
<CodeMirror
  options={
    Object {
      "gutters": Array [
        "annotations",
        "CodeMirror-linenumbers",
      ],
      "lineNumbers": true,
      "lineWrapping": true,
      "readOnly": true,
    }
  }
  preserveScrollPosition={false}
  value="Wubba lubba Dub Dub"
/>
`;
